src/v/raft/heartbeat_manager.cc

```diff
@@ -125,16 +127,138 @@ heartbeat_manager::heartbeat_requests heartbeat_manager::requests_for_range() {
+                // here as those fields are not used to process lightweight
+                // heartbeats
+                it->second.emplace_back(
+                  group_beat,
+                  heartbeat_manager::follower_request_meta(
+                    r, raft::follower_req_seq{}, model::offset{}, id));
+                continue;
+            }
+            r->_probe->full_heartbeat();
+            auto const seq_id = ++follower_metadata.last_sent_seq;
+
+            follower_metadata.last_sent_protocol_meta = raft_metadata;
+            group_beat.data = heartbeat_request_data{
+              .source_revision = r->_self.revision(),
+              .target_revision = id.revision(),
+              .commit_index = raft_metadata.commit_index,
+              .term = raft_metadata.term,
+              .prev_log_index = raft_metadata.prev_log_index,
+              .prev_log_term = raft_metadata.prev_log_term,
+              .last_visible_index = raft_metadata.last_visible_index,
+            };
+            it->second.emplace_back(
+              group_beat,
+              heartbeat_manager::follower_request_meta(
+                r, seq_id, raft_metadata.prev_log_index, id));
+
+            if (r->should_reconnect_follower(follower_metadata)) {
+                reconnect_nodes.insert(id.id());
+            }
+        }
+    }
+
+    std::vector<heartbeat_manager::node_heartbeat_v2> reqs;
+    reqs.reserve(pending_beats.size());
+    for (auto& p : pending_beats) {
+        ss::chunked_fifo<group_heartbeat> requests;
+        absl::node_hash_map<
+          raft::group_id,
+          heartbeat_manager::follower_request_meta>
+          meta_map;
+        requests.reserve(p.second.size());
+        heartbeat_request_v2 req(_self, p.first);
+        for (auto& [hb, follower_meta] : p.second) {
+            meta_map.emplace(hb.group, std::move(follower_meta));
+            req.add(hb);
+        }
+        reqs.emplace_back(p.first, std::move(req), std::move(meta_map));
+    }
+
+    return heartbeat_requests_v2{
+      .requests{std::move(reqs)}, .reconnect_nodes{reconnect_nodes}};
+}
+
+bool heartbeat_manager::needs_full_heartbeat(
+  const follower_index_metadata& f_meta,
+  const protocol_metadata& p_meta,
+  model::offset leader_flushed_offset) const {
+    /**
+     * This condition makes sending lw_heartbeats not vulnerable for
+     * requests/replies reordering.
+     *
+     * We only send lw_heartbeat if last received not reordered reply was
+     * successful and follower acknowledged state is equal to current leader
+     * state.
+     *
+     * Full heartbeat will be sent to the follower every time its responded with
+     * error, requests were reordered or leader log was flushed.
+     */
+
+    return f_meta.last_sent_seq != f_meta.last_successful_received_seq
+           || f_meta.last_sent_protocol_meta != p_meta
+           || leader_flushed_offset != f_meta.last_flushed_log_index;
```

Comment:
can you explain the last condition a bit more? why it warrants a full hb?

Reply:
i will add a code comment